### PR TITLE
Suppress execution context from token renewal

### PR DESCRIPTION
**Pull request by jviau**
Suppress execution context from the background token renewal operations. The renewal uses the logger, and so we will capture whatever current log scope is present, leaking that scope into this background thread. This is both a small memory leak and potential object disposed failure (the log scope or its contents may be disposed at any point).

### Issue describing the changes in this PR

resolves #issue_for_this_pr

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)
* [x] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk
* [x] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
    * [ ] Otherwise: major or minor version updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`
* [x] My changes **do not** add EventIds to our EventSource logs
    * [ ] Otherwise: Ensure the EventIds are within the supported range in our existing Windows infrastructure. You may validate this with a deployed app's telemetry. You may also extend the range by completing a PR such as [this one](https://msazure.visualstudio.com/One/_git/AAPT-Antares-Websites/pullrequest/7463263?_a=files).
* [x] My changes **should** be added to v3.x branch.
    * [ ] Otherwise: This change only applies to Durable Functions v2.x and **will not** be merged to branch v3.x.

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/Auth/AzureCredentialFactory.cs [hunk 1]
```diff
@@ -72,6 +72,10 @@ public AppAuthTokenCredential Create(IConfiguration configuration, TimeSpan toke
                 RefreshOffset = tokenRefreshOffset,
             };
 
+            // The token credential will make background callbacks to renew the token.
+            // We suppress async flow to avoid logging scope from being captured as we do not know
+            // where this will be called from first.
+            using AsyncFlowControl flowControl = System.Threading.ExecutionContext.SuppressFlow();
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Auth/AzureCredentialFactory.cs:78]: What logging scope is currently being captured, and why does it need to be suppressed? I'm missing a bit of background context on this. Just wondering how confident we feel that this fix will be effective.

Comment (jviau) [reply src/WebJobs.Extensions.DurableTask/Auth/AzureCredentialFactory.cs:78]: We use `ILogger` in the trace helper here. The log scope of whatever was present before calling `.Create` is captured.

I did manual tests with an `AsyncLocal` and verified before the value set outside is present during renew, and after the change it is no longer present.

Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Auth/AzureCredentialFactory.cs:78]: Makes sense at a high level. The thing I can't wrap my head around is that the thing we captured appears to have a lifetime that's different from this AzureCredentialFactory (a singleton). How is that possible? Shouldn't singletons only capture context that also has a singleton lifecycle? For this particular crash, it appears that a singleton object captured a per-request object, which seems very concerning to me.

**Comment by jviau** [comment]
I wonder if this will address https://github.com/Azure/azure-functions-durable-extension/issues/2771?

File: src/WebJobs.Extensions.DurableTask/Auth/AzureCredentialFactory.cs (hunk 1 above)
Comment (jviau) [reply src/WebJobs.Extensions.DurableTask/Auth/AzureCredentialFactory.cs:78]: Yes, a singleton did capture a per-request object! But it didn't capture it through DI (nor at time of construction). It captured it through execution context and at the time/where `.Create` was called. Async locals are very powerful, but also lead to subtle issues like this.

So yes, it is a singleton, but the 'scope' I talk about in the PR is not related to service scope directly.

Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Auth/AzureCredentialFactory.cs:78]: Thanks for the clarification. I think I understand it now. Makes me much more wary of async locals as silent app killers. 😅

**Review by cgillum — APPROVED** [review]

**Pull request merged**